sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -501,6 +502,42 @@ object RemoveNoopOperators extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Remove no-op `Union` from the query plan that do not make any modifications.
+ */
+object RemoveNoopUnion extends Rule[LogicalPlan] {
+  private def removeAliasOnlyProject(plan: LogicalPlan): LogicalPlan = plan match {
+    case p @ Project(projectList, child) =>
+      val originalOutputs = projectList.collect {
```

Comment:
I feel `collect` is kind of dangerous here, we essentially ignore all expressions that are not attribute or alias of attribute here. How can it work correctly?

I am thinking of example query like

```
SELECT DISTINCT k, expr
FROM
(
  SELECT k, k + 1 AS expr
  FROM test_table
  UNION ALL
  SELECT k, k + 2 AS expr
  FROM test_table
)
```

Reply:
@viirya  Shall we have a test case suggested by @c21 suggested?

Reply:
Sure.

Reply:
In addition to adding tests, can we also add comments to explain it? `Project(a, a + 1 as res, Table(a))` is not alias-only but seems the code here will remove the `Project`.

Reply:
I misunderstood the point. This kind of `Project` should not be removed, actually.